src/misc/degree_sequence.cpp

```diff
@@ -68,66 +70,232 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
 /***** Undirected simple graphs ******/
 /*************************************/
 
-// Generate simple undirected realization as edge-list.
-// If largest=true, always choose the vertex with the largest remaining degree to connect up next.
-// Otherwise, always choose the one with the smallest remaining degree.
-static igraph_error_t igraph_i_havel_hakimi(const igraph_vector_int_t *deg, igraph_vector_int_t *edges, bool largest) {
-    igraph_integer_t n = igraph_vector_int_size(deg);
+// "Bucket Node" for nodes of the same degree
+struct BNode {
+    igraph_integer_t count = 0;
+    std::stack<vd_pair> nodes;
+    igraph_integer_t next;       // next bucket (higher degree)
```

Comment:
Style: Please separate line comments from the end of the code by only a single space.